### ISSUE TITLE
`:line-length` does not honor `:clj-kondo/ignore`

### ISSUE THREAD

**Issue by colin-p-hill** [clj-kondo/clj-kondo#1952]
**version**

v2023.01.12

**platform**

Windows 11, native, scoop.

**editor**

Cursive, clojure-extras.

**problem**

The `:line-length` linter reports findings for expressions that are annotated with `#_:clj-kondo/ignore` or `#_{:clj-kondo/ignore [:line-length]}`.

**repro**

```clojure
(ns repro)
#_:clj-kondo/ignore
;; This is a very long comment aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
#_:clj-kondo/ignore
(def this-is-a-very-long-symbol-aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa nil)
```
```shell
> clj-kondo --lint repro.clj
repro.clj:3:81: warning: Line is longer than 80 characters.
repro.clj:5:81: warning: Line is longer than 80 characters.
linting took 7ms, errors: 0, warnings: 2
```

**expected behavior**

Expressions annotated to ignore this linter do not trigger findings.

**Comment by borkdude** [comment]
Sometimes it's just impossible to get something below the line length.

I've thought about changing line-length to only complain about new tokens that start after the max line length. I'm not sure if that would be a good solution.  What do you think of that? 

Agree that clj-kondo/ignore should work.

**Comment by colin-p-hill** [comment]
The original motivating scenario for this bug report is omitted from the report, because it's probably even trickier: Comments containing URLs, which exceed the length limit but cannot reasonably be wrapped. Sample from one of my projects:

```clojure
   ;; Column 80 marked at the vertical bar:                                    |
   ;; WARNING: `s/nonconforming` is undocumented. It is unlikely to disappear
   ;;          from spec1, but may not exist in spec2.
   ;; 2017, Alex says Rich is unsure about whether to include it:
   ;; https://clojurians-log.clojureverse.org/clojure-spec/2017-08-12/1502573905.650871
   ;; 2022, Alex suggests using it to someone:
   ;; https://clojurians-log.clojureverse.org/beginners/2022-12-22/1671717166.291839
   ;; Note that this doesn't seem to be the use case Alex has in mind.
   ;; If needed, we can mimic it by calling `s/valid?` in a lambda.
```

Not sure whether your reader can handle ignore annotations on comments, but here's a use case for doing so.

**Comment by borkdude** [comment]
This happened to me too once with URLs. Comments are processed differently than real code, so a full repro of this would be good since it would not have been fixed otherwise.

**Comment by colin-p-hill** [comment]
Updated. The repro file differs slightly from that of #1951 because it turns out clj-kondo chokes on an annotation preceding a comment at the end of the file:
```clojure
(ns repro)
#_:clj-kondo/ignore
;; This is a very long comment aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
```
```shell
> clj-kondo --lint repro.clj
repro.clj:0:0: error: Can't parse repro.clj, clojure.lang.PersistentList cannot be cast to clojure.lang.IPersistentMap
linting took 4ms, errors: 1, warnings: 0
```
Distinct manifestation but seems likely related to the ability to annotate comments at all. Do you want a separate report for that, or should I fold that into this one as well?

**Comment by borkdude** [comment]
This is expected since `#_` must always be followed by a form

**Comment by borkdude** [comment]
No sorry, that doesn't make sense, the keyword is already the form. Yeah, report welcome

**Comment by colin-p-hill** [comment]
Reported as #1953

**Comment by borkdude** [comment]
So `#_clj-kondo/ignore` refers to the next form, but comments aren't forms...

It would be good to check how other linters (rubocop, eslint) allow users to suppress lint warnings in these situations.

**Comment by borkdude** [comment]
Rubocop does it like this:

```
# rubocop:disable Layout/LineLength
  "I'm a really long line"
# rubocop:enable Layout/LineLength
```

**Comment by borkdude** [comment]
I think as a first step, we could be lenient if a token occurs in the offensive line that is longer than the limit, this would likely solve your issue, right?

**Comment by borkdude** [comment]
+ configurable offset, like you said.

**Comment by borkdude** [comment]
also some interesting options here:

https://stackoverflow.com/questions/27732209/turning-off-eslint-rule-for-a-specific-line/29592334#29592334

e.g. 

- `#_:clj-kondo/ignore-line`
- `#_:clj-kondo/ignore-next-line`

or:

``` clojure
;; clj-kondo/disable
;; no warnings here
;; clj-kondo/enable
```

I feel I must think a little bit harder about this before settling on a solution...

Perhaps the quickest solution for now is `:ignore-urls true` and if a line matches `http` we ignore it.

**Comment by NoahTheDuke** [comment]
Rubocop works by having ast nodes for comments (as seen [here](https://github.com/rubocop/rubocop-ast/blob/8a5e1679a42cfd28d086c355b2ae263ada47cb6b/lib/rubocop/ast/token.rb#L57)). The [`Layout/LineLength`](https://github.com/rubocop/rubocop/blob/8698ae90837424747e9ce00a7596346ae2509ed1/lib/rubocop/cop/layout/line_length.rb) cop relies on a `processed_source` that holds an iterator of `lines`, and is able to step over those lines and [checks each for length and directives](https://github.com/rubocop/rubocop/blob/8698ae90837424747e9ce00a7596346ae2509ed1/lib/rubocop/cop/layout/line_length.rb#L166-L168).

clj-kondo doesn't track comments or have a concept of "lines" the same way, right?

**Comment by borkdude** [comment]
clj-kondo strips out all whitespace before it processes the code (since initially I didn't expect to write linters for whitespace).

I'll add `:exclude-urls` and `:exclude-pattern`,  I think that will cover the majority of cases for now.

``` clojure
(deftest exclusions-test
  (is (empty? (lint! " ;; https://clojurians-log.clojureverse.org/clojure-spec/2017-08-12/1502573905.650871"
                     '{:linters {:line-length {:exclude-urls true
                                               :level :warning
                                               :max-line-length 80}}})))
  (is (empty? (lint! " ;; :ll/ok"
                     '{:linters {:line-length {:exclude-pattern ";; :ll/ok"
                                               :level :warning}}}))))
```

**Comment by colin-p-hill** [comment]
I think the URL exclusion will cover every place I've run into this in actual practice.
